File: src/Binary_Search_Tree_sort/BST.java
package Binary_Search_Tree_sort;

public class BST {
    Node root;

    BST() {
        root = null;
    }

    void insert(int key) {
        root = insertRec(root, key);
    }

    Node insertRec(Node root, int key) {
        if (root == null) {
            return new Node(key);
        }

        if (key < root.key) {
            root.left = insertRec(root.left, key);
        } else if (key > root.key) {
            root.right = insertRec(root.right, key);
        }

        return root;
    }

    void inOrderTraversal(Node root, int[] sortedArray, int[] index) {
        if (root != null) {
            inOrderTraversal(root.left, sortedArray, index);
            sortedArray[index[0]++] = root.key;
            inOrderTraversal(root.right, sortedArray, index);
        }
    }

    void bstSort(int[] arr) {
        for (int num : arr) {
            insert(num);
        }

        int[] index = { 0 };
        inOrderTraversal(root, arr, index);
    }

    static void printArray(int[] arr) {
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int[] arr = { 50, 30, 70, 20, 40, 60, 80 };
        System.out.println("before the sort :");
        printArray(arr);

        BST tree = new BST();
        tree.bstSort(arr);

        System.out.println("after the sort :");
        printArray(arr);
    }
}

File: Sortingc++/BST.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    Node *left;
    Node *right;
};
Node *createNode(int value)
{
    Node *newNode = new Node();
    newNode->data = value;
    newNode->left = newNode->right = nullptr;
    return newNode;
}

Node *insert(Node *root, int value)
{
    if (root == nullptr)
    {
        return createNode(value);
    }
    if (value < root->data)
    {
        root->left = insert(root->left, value);
    }
    else
    {
        root->right = insert(root->right, value);
    }
    return root;
}
void inorder(Node *root)
{
    if (root != nullptr)
    {
        inorder(root->left);
        cout << root->data << " ";
        inorder(root->right);
    }
}

int main()
{
    Node *root = nullptr;
    int a[] = {5, 3, 7, 2, 8, 1, 6};
    for (int i = 0; i < 7; i++)
    {
        root = insert(root, a[i]);
    }
    inorder(root);
    return 0;
}